### DISCUSSION TITLE
create SLP project from scratch using python

### DISCUSSION THREAD

**Discussion by deemok7** [talmolab/sleap#1534] [Help!] [answered]
I have keypoint of each node for each frame of a video, I need a code like below to create an SLEAP project.
```python
import sleap
from sleap import Labels, Video, LabeledFrame, Instance, Skeleton
from sleap.skeleton import Node

# Create a new SLEAP project
labels = Labels()

# Create a video object and add it to the project
video = Video.from_filename("video.mp4")
labels.add_video(video)

# Create a labeled frame and add it to the project
frame = LabeledFrame(video=video, frame_idx=0)
labels.frames.append(frame)

# Create a skeleton and add it to the project
skeleton = Skeleton()
skeleton.add_node(Node("head"))
skeleton.add_node(Node("tail"))
labels.add_skeleton(skeleton)

# Create an instance and add it to the labeled frame
instance = Instance(skeleton=skeleton)
instance.add_node("head", (100, 100))
instance.add_node("tail", (200, 200))
frame.add_instance(instance)

# Save the project to a SLEAP file
sleap.save_file(labels, "labels.slp")
```

**Comment by roomrys** [comment, marked as answer]
Hi @mmmahdiii,
 
 When creating a `Labels` object, create it from the inside out by creating object that are referenced by other objects first:
 ```python
import sleap
from sleap import Labels, Video, LabeledFrame, Instance, Skeleton
from sleap.instance import Point
from sleap.skeleton import Node

 # Create a video object
video_path = r"path/to/video.mp4"
video = Video.from_filename(video_path)
 
# Create a skeleton
nodes = ["head", "tail"]
skeleton = Skeleton(name="my_skeleton")
skeleton.add_nodes(nodes)
skeleton.add_edge("head", "tail")

# Initialize a list of labeled frames
lfs = []

# 0. Create a list of instances for this labeled frame
instances = []

# 1. Create an instance and append to list of instances
points = dict(head=Point(100, 100), tail=Point(200, 200))
inst = Instance(skeleton=skeleton, points=points)
instances.append(inst)

# 2. Create a labeled frame and add it to the list
frame = LabeledFrame(video=video, frame_idx=0, instances=instances)
lfs.append(frame)

# Repeat steps 0-2 for all frames you would like to add (preferably in a for-loop)
...

# Create the Labels object and save to an slp
ds = r"path/to/save/labels.slp"
labels = Labels(lfs)
labels.save(ds)
```

Thanks,
Liezl

**Reply by isabellow** [reply]
Hi @roomrys,

I'm using this approach to make a SLEAP project using labeled frames from different GUI. I think it's largely working well--thank you!

I'm running into issues with the video object. My labeling GUI outputs video frames as a numpy array called `images` of shape `(num_frames, width, height, 1)`. So I am using `video = Video.from_numpy(images)` to create the video object (instead of `Video.from_filename(video_path)` as above). This seems to work fine for labeling the frames (or at least it is not throwing any errors). But when I try to open the project in the labeling GUI, it says that the raw video data cannot be found.

What is the best way to save my video to the SLEAP project? Please let me know if you need more information or code snippets.

Thanks,
Isabel

**Reply by talmo** [reply]
Hi @isabellow,

Numpy videos are not very well supported since we don't store them on disk or in the SLP file. Do you mind [opening a new discussion](https://github.com/talmolab/sleap/discussions/new/choose) so we can follow up there if you're still running into issues?

In general, our standalone package [`sleap-io`](https://io.sleap.ai/latest/) is also great for doing things like programmatically creating SLEAP projects.

Cheers,

Talmo

**Reply by isabellow** [reply]
Hi Talmo,

Thanks for the reply! I realized that `Video.from_filename(video_path)` will accept a path to a numpy array. From there, everything worked great. I have some preliminary tracking data that's not bad so far!

Best,
Isabel

**Comment by deemok7** [comment]
Hi @roomrys, 
it works perfectly. You always helped me. 

Thanks a million!
Mahdi
